src/test/groovy/graphql/util/TraverserTest.groovy

```diff
@@ -327,10 +347,106 @@ class TraverserTest extends Specification {
 
+    def "test node position"() {
+        given:
+        def visitor = [
+                enter: { TraverserContext context ->
+                    def curAcc = context.getCurrentAccumulate()
+                    if (context.getPosition() != null) {
+                        curAcc.add(context.getPosition().index)
+                    }
+                    TraversalControl.CONTINUE
+                },
+                leave: { TraverserContext context ->
+                    TraversalControl.CONTINUE
+                }
+        ] as TraverserVisitor
+        when:
+        def result = Traverser.depthFirst({ n -> n.children }, null, [])
+                .traverse(root, visitor)
+
+
+        then:
+        result.accumulatedResult == [0, 0, 1, 0, 1]
+    }
+
+    def treeWithNamedChildren() {
+        def leaf1 = [number: 3, children: [:]]
+        def leaf2 = [number: 4, children: [:]]
+        def leaf3 = [number: 5, children: [:]]
+        def leaf4 = [number: 6, children: [:]]
+
+        def node1 = [number: 1, children: [a: [leaf1], b: [leaf2]]]
+        def node2 = [number: 2, children: [c: [leaf3], d: [leaf4]]]
+
+        [number: 0, children: [x: [node1], y: [node2]]]
+    }
+
+    def "breadth first with named children"() {
+        def root = treeWithNamedChildren()
+        given:
+        def visitor = [
+                enter: { TraverserContext context ->
+                    def curAcc = context.getCurrentAccumulate()
+                    if (context.getPosition() != null) {
+                        curAcc.add(context.getPosition())
+                    }
+                    TraversalControl.CONTINUE
+                },
+                leave: { TraverserContext context ->
+                    TraversalControl.CONTINUE
+                }
+        ] as TraverserVisitor
+        when:
+        def result = Traverser.breadthFirstWithNamedChildren({ n -> n.children }, null, [])
+                .traverse(root, visitor)
+
+
+        then:
+        result.accumulatedResult == [new NodePosition("x", 0),
+                                     new NodePosition("y", 0),
+                                     new NodePosition("a", 0),
+                                     new NodePosition("b", 0),
+                                     new NodePosition("c", 0),
+                                     new NodePosition("d", 0),
+        ]
+    }
+
+    def "depth first with named children"() {
+        def root = treeWithNamedChildren()
+        given:
+        def visitor = [
+                enter: { TraverserContext context ->
+                    def curAcc = context.getCurrentAccumulate()
+                    if (context.getPosition() != null) {
+                        curAcc.add(context.getPosition())
+                    }
+                    TraversalControl.CONTINUE
+                },
+                leave: { TraverserContext context ->
+                    TraversalControl.CONTINUE
+                }
+        ] as TraverserVisitor
```

Comment:
what is this crazy syntax?? I didnt know you could do this! :)

Reply:
a groovy map with closures/functions as value ... not so crazy, or?